### PR TITLE
Feature/hibernate integration

### PR DESCRIPTION
**Pull request by 8r1m2t0n3**

### CODE REVIEW

File: src/main/java/org/andersen/homework/util/SessionFactoryManager.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import org.hibernate.cfg.Configuration;
+
+public class SessionFactoryManager {
+
+  @Getter
+  private static final SessionFactory sessionFactory = buildSessionFactory();
+
+  private static SessionFactory buildSessionFactory() {
+    try {
+      return new Configuration().configure("/hibernate.cfg.xml").buildSessionFactory();
```
Comment (Darya-Drutsko) [inline src/main/java/org/andersen/homework/util/SessionFactoryManager.java]: do we actually need hardcoded value here?

File: src/main/java/org/andersen/homework/service/TicketService.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+  public Optional<Ticket> getByUserId(UUID userId) {
+    return Optional.ofNullable(ticketDao.findByUserId(userId));
+  }
+
+  public List<Ticket> getAll() {
+    return ticketDao.getAll();
+  }
+
+  public static ConcertTicket getRandomConcertTicket() {
```
Comment (Darya-Drutsko) [inline src/main/java/org/andersen/homework/service/TicketService.java]: separate this to new util class

File: src/main/java/org/andersen/homework/service/TicketService.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        BigDecimal.valueOf(((float) RandomizerUtil.getRandomInt(10, 1000)) / 100),
+        RandomizerUtil.getRandomString(RandomizerUtil.getRandomInt(1, 10)),
+        (short) RandomizerUtil.getRandomInt(0, 999),
+        ((float) RandomizerUtil.getRandomInt(10, 1000)) / 100,
+        RandomizerUtil.getRandomFromEnum(StadiumSector.class),
+        LocalDateTime.now(),
+        RandomizerUtil.getRandomInt(0, 1) == 1 ? Boolean.TRUE : Boolean.FALSE);
+  }
+
+  public static BusTicket getRandomBusTicket() {
```
Comment (Darya-Drutsko) [inline src/main/java/org/andersen/homework/service/TicketService.java]: separate this to util class

File: src/main/java/org/andersen/homework/model/entity/user/Client.java [hunk 4]
```diff
@@ -1,26 +1,27 @@
...
+@Entity
+@DiscriminatorValue("CLIENT")
 @Getter
 @Setter
+@NoArgsConstructor
 @ToString(callSuper = true)
 @EqualsAndHashCode(callSuper = true)
 public class Client extends User {
 
+  @OneToOne(cascade = CascadeType.REMOVE)
```
Comment (Darya-Drutsko) [inline src/main/java/org/andersen/homework/model/entity/user/Client.java:24]: Why OneToOne relation?

File: src/main/java/org/andersen/homework/model/entity/ticket/Ticket.java [hunk 5]
```diff
@@ -1,33 +1,56 @@
...
 import lombok.Getter;
 import lombok.NoArgsConstructor;
 import lombok.Setter;
-import lombok.ToString;
-import org.andersen.homework.annotation.NullableWarning;
+import org.andersen.homework.model.entity.user.Client;
 
+@Entity
+@Table(name = "ticket")
+@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
```
Comment (Darya-Drutsko) [inline src/main/java/org/andersen/homework/model/entity/ticket/Ticket.java:25]: Why do we need here Inheritance annotation? Why exact this InheritanceType?

File: src/main/java/org/andersen/homework/model/entity/user/User.java [hunk 6]
```diff
@@ -1,20 +1,35 @@
...
 import lombok.EqualsAndHashCode;
 import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
 import lombok.ToString;
-import org.andersen.homework.model.enums.UserRole;
 
+@Entity
+@Table(name = "\"user\"")
+@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
```
Comment (Darya-Drutsko) [inline src/main/java/org/andersen/homework/model/entity/user/User.java:22]: Why have you chosen this InheritanceType?

File: src/main/java/org/andersen/homework/model/entity/user/Client.java (hunk 4 above)
Comment (8r1m2t0n3) [reply src/main/java/org/andersen/homework/model/entity/user/Client.java:24]: One client may have only one ticket, same for ticket. It wasn't specified in the task to make ManyToOne relation so a made it OneToOne. Do i have to change it?

File: src/main/java/org/andersen/homework/model/entity/ticket/Ticket.java (hunk 5 above)
Comment (8r1m2t0n3) [reply src/main/java/org/andersen/homework/model/entity/ticket/Ticket.java:25]: I have Ticket as parent class, ConcertTicket and BusTicket as children, inheriting the Ticket. `@Inheritance` allows to set the way how to store parents and children in db. With `InheritanceType.SINGLE_TABLE` all the inheritors of the Ticket will be stored in the same table. A discriminator column called `type`, will be used to distinguish between the different subclasses.

File: src/main/java/org/andersen/homework/model/entity/user/User.java (hunk 6 above)
Comment (8r1m2t0n3) [reply src/main/java/org/andersen/homework/model/entity/user/User.java:22]: It's a simple way to store parents and children without creating extra tables.

File: src/main/java/org/andersen/homework/util/SessionFactoryManager.java (hunk 1 above)
Comment (8r1m2t0n3) [reply src/main/java/org/andersen/homework/util/SessionFactoryManager.java]: No, redundant. Will remove it

File: src/main/java/org/andersen/homework/model/entity/user/Client.java (hunk 4 above)
Comment (Darya-Drutsko) [reply src/main/java/org/andersen/homework/model/entity/user/Client.java:24]: No, it's not really matters. Just asking why have you chose it)
In real life seems that each user could have multiple tickets, but if you say that for your project requirements only one - fine for me also :)

File: src/main/java/org/andersen/homework/model/entity/ticket/Ticket.java (hunk 5 above)
Comment (Darya-Drutsko) [reply src/main/java/org/andersen/homework/model/entity/ticket/Ticket.java:25]: Cool, also would be great to read/learn about other inheritance types and get perspective in which cases need to choose this or that type

**Review by Darya-Drutsko — APPROVED** [review]

**Pull request merged**